### PR TITLE
type registration for serializers

### PR DESCRIPTION
**Pull request by Carlm-MS**
- bug fix for powershell installer
- scan types for use with external serializers

### CODE REVIEW

File: src/Orleans/Serialization/SerializationManager.cs [hunk 1]
```diff
@@ -2280,6 +2280,23 @@ private static void RegisterSerializationProviders(List<TypeInfo> providerTypes)
...
+                        type => 
+                        (type.IsClass || type.IsValueType) && !type.IsAbstract &&
+                        externalSerializers.Any(ser => ser.IsSupportedType(type))))
+                    {
+                        SerializationManager.Register(type);
+                    }
+                };
+
+            AppDomain.CurrentDomain.AssemblyLoad += (o, e) => registerAction(e.LoadedAssembly);
+            Array.ForEach(AppDomain.CurrentDomain.GetAssemblies(), registerAction);
```
Comment (jdom) [inline src/Orleans/Serialization/SerializationManager.cs]: This seems useful and might be related to some serialization issues we have been seeing in our tests, that prompt use to do workarounds such as calling `SerializationManager.InitializeForTesting()` before starting the cluster (although I'm not entirely sure if it's the same root cause).
Nevertheless this seems overly aggressive adding every single type that is loaded and can be serialized... can you or someone think of an approach that is a little bit more lazy than this?

Comment (Carlm-MS) [inline src/Orleans/Serialization/SerializationManager.cs]: The only way of doing this is to add some methods to the TypeUtilities class that allows you to resolve the type from the OrleansTypeKeyString.  Currently, it's just a dictionary lookup which is why the types need to be registered prior to deserialization.

Comment (Carlm-MS) [inline src/Orleans/Serialization/SerializationManager.cs]: In order to resolve the types from the type key string, there are a few annoyances:

- All types in namespaces that start with System are truncated, so to resolve types from the system namespace, you'd have to enumerate through all of the system namespaces in all of the assemblies and see if that type exists
- nested types are delimited using the period character which means that for a type key string like Mine.Orleans.Serialization, it is ambiguous as it is the type Mine+Orleans+Serialization in the global namespace, Orleans+Serialization in the Mine namespace, or Serialization in the Mine.Orleans namespace.  All of these combos would have to be evaluated against all loaded assemblies

Comment (Carlm-MS) [inline src/Orleans/Serialization/SerializationManager.cs]: @jdom Actually, a less brute force mechanism would be to scan all assemblies for grain interfaces and then find all method parameters/return value types.  Then register those. 

Can you think of any issues with that approach?

**Comment by sergeybykov** [comment]
This one seems to have broken many tests. I'll rerun to confirm.

**Comment by Carlm-MS** [comment]
The technique I through of in my previous comment will not work for generic grains since the types can't be determined by examining the method info, so scratch that idea.

Another way I'm investigating right now is to add a check in the GrainFactory.GetGrain and StreamProvider.GetStream methods to see if the types have been registered.

**Comment by Carlm-MS** [comment]
@sergeybykov I modified the code to be lazily discover types that haven't been registered by doing a one-time scan of the IGrain type methods/generic parameters when GetGrain is called.  It should be less likely to cause any problems with tests.

**Comment by sergeybykov** [comment]
@Carlm-MS I kicked off functional tests for the latest version.

File: src/Orleans/Core/GrainFactory.cs [hunk 2]
```diff
@@ -167,6 +179,50 @@ internal Task<TGrainObserverInterface> CreateObjectReference<TGrainObserverInter
...
+                                {
+                                    return parameterTypes.Concat(mi.ReturnType.GetGenericArguments());
+                                }
+
+                                return parameterTypes.Concat(new[] { mi.ReturnType });
+                            })
+                        .Distinct()
+                        .Where(SerializationManager.HasExternalSerializer))
+                    {
+                        SerializationManager.Register(paramType);
```
Comment (jdom) [inline src/Orleans/Core/GrainFactory.cs]: This approach wouldn't take derived types into account... Not sure what's the best approach... @sergeybykov what do you think we should do?

Comment (Carlm-MS) [inline src/Orleans/Core/GrainFactory.cs]: After more investigation, the Assembly Processor does a brute force scan of all assemblies that get loaded to find serialization info.  A minor change to that logic should be all that is necessary to fix this and handle the derived types issue as well.

Sorry for the churn.



The current mechanism is a bit verbose as well.

**Comment by Carlm-MS** [comment]
Found a better location to implement the logic.  The AssemblyProcessor already does a brute force scan of all type definitions in all assemblies.  Adding a tiny amount of code should do what we want.

Sorry for the churn.

**Comment by sergeybykov** [comment]
@Carlm-MS I kicked off another test.

**Comment by veikkoeeva** [comment]
Should some [PowerShell headers](https://technet.microsoft.com/en-us/library/hh847765.aspx) be added too? I mean specifically something like `#Requires -version 3.0`, `#Requires -RunAsAdministrator` (if needed) etc. Some [more explanation](https://technet.microsoft.com/en-us/magazine/ff458353.aspx).

**Comment by Carlm-MS** [comment]
@veikkoeeva In this case, I don't think it's necessary since it's just a script that launches for the NuGet package installer. It's not intended for use in any other context.

**Comment by veikkoeeva** [comment]
@Carlm-MS Good point.

**Comment by sergeybykov** [comment]
Tests ran fine. Merging it. Thank you, @Carlm-MS!

**Pull request merged**